contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -114,19 +121,38 @@ contract KeystoneForwarder is OwnerIsCreator, ITypeAndVersion, IRouter {
     bytes calldata metadata,
     bytes calldata validatedReport
   ) public returns (bool) {
-    if (!s_forwarders[msg.sender]) {
-      revert UnauthorizedForwarder();
-    }
+    if (!s_forwarders[msg.sender]) revert UnauthorizedForwarder();
+    uint256 gasLeft = gasleft();
+    if (gasLeft < REQUIRED_GAS_FOR_ROUTING) revert InsufficientGasForRouting(transmissionId);
 
-    if (s_transmissions[transmissionId].transmitter != address(0)) revert AlreadyAttempted(transmissionId);
+    Transmission memory transmission = s_transmissions[transmissionId];
+    if (transmission.success || transmission.invalidReceiver) revert AlreadyAttempted(transmissionId);
+
+    uint256 gasLimit = gasLeft - REQUIRED_GAS_FOR_ROUTING;
     s_transmissions[transmissionId].transmitter = transmitter;
+    s_transmissions[transmissionId].gasLimit = uint80(gasLimit);
+
+    if (receiver.code.length == 0) {
+      s_transmissions[transmissionId].invalidReceiver = true;
+      return false;
+    }
 
-    if (receiver.code.length == 0) return false;
+    try IERC165(receiver).supportsInterface(type(IReceiver).interfaceId) {
+      bool success;
+      bytes memory payload = abi.encodeCall(IReceiver.onReport, (metadata, validatedReport));
 
-    try IReceiver(receiver).onReport(metadata, validatedReport) {
-      s_transmissions[transmissionId].success = true;
-      return true;
+      assembly {
+        // call and return whether we succeeded. ignore return data
+        // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+        success := call(gasLimit, receiver, 0, add(payload, 0x20), mload(payload), 0x0, 0x0)
```

Comment:
That success == false mean any reason for revert or only out of gas?

Reply:
It can be either a revert (for whatever user reason) or out of gas (technically, still a revert).